src/runtime/openglcompute.cpp

```diff
@@ -0,0 +1,608 @@
+
+    size_t size = buf_size(user_context, buf);
+
+    if (buf->dev) {
+        // This buffer already has a device allocation
+        debug(user_context) << "openglcompute_device_malloc: This buffer already has a "
+                               "device allocation\n";
+        return 0;
+    }
+
+    halide_assert(user_context, buf->stride[0] >= 0 && buf->stride[1] >= 0 &&
+                                buf->stride[2] >= 0 && buf->stride[3] >= 0);
+
+    debug(user_context) << "    allocating "
+                        << " buffer of " << (int64_t)size << " bytes, "
+                        << "extents: " << buf->extent[0] << "x"
+                        << buf->extent[1] << "x" << buf->extent[2] << "x"
+                        << buf->extent[3] << " "
+                        << "strides: " << buf->stride[0] << "x"
+                        << buf->stride[1] << "x" << buf->stride[2] << "x"
+                        << buf->stride[3] << " "
+                        << "(" << buf->elem_size << " bytes per element)\n";
+
+    if (int error = halide_openglcompute_init(user_context)) {
+        return error;
+    }
+    debug(user_context) << "openglcompute_device_malloc: initialization completed.\n";
+
+    if (!buf) {
+        error(user_context) << "Invalid buffer";
+        return 1;
+    }
+
+    GLuint the_buffer;
+    global_state.GenBuffers(1, &the_buffer);
+    if (global_state.CheckAndReportError(user_context, "oglc: GenBuffers")) { return 1; }
+    global_state.BindBuffer(GL_ARRAY_BUFFER, the_buffer);
+    if (global_state.CheckAndReportError(user_context, "oglc: BindBuffer")) { return 1; }
+    size_t sizeInBytes = buf_size(user_context, buf);
+    global_state.BufferData(GL_ARRAY_BUFFER, sizeInBytes, NULL, GL_DYNAMIC_COPY);
+    if (global_state.CheckAndReportError(user_context, "oglc: BufferData")) { return 1; }
+
+    buf->dev = halide_new_device_wrapper(the_buffer, &openglcompute_device_interface);
+
+    debug(user_context) << "Allocated dev_buffer(i.e. vbo) " << the_buffer << "\n";
+
+#ifdef DEBUG_RUNTIME
+    uint64_t t_after = halide_current_time_ns(user_context);
+    debug(user_context) << "    Time: " << (t_after - t_before) / 1.0e6
+                        << " ms for malloc\n";
+#endif
+
+    return 0;
+}
+
+WEAK int halide_openglcompute_device_free(void *user_context, buffer_t *buf) {
+#ifdef DEBUG_RUNTIME
+    uint64_t t_before = halide_current_time_ns(user_context);
+#endif
+
+    if (!global_state.initialized) {
+        error(user_context) << "OpenGL runtime not initialized in call to halide_openglcompute_device_free.";
+        return 1;
+    }
+
+    if (buf->dev == 0) {
+        return 0;
+    }
+
+    GLuint the_buffer = halide_get_device_handle(buf->dev);
+    debug(user_context) << "OGLC: halide_openglcompute_device_free ("
+                        << "user_context: " << user_context
+                        << ", the_buffer:" << the_buffer
+                        << ")\n";
+
+    // TODO(aam): implement this
```

Comment:
Thank you! I'm glad you reminded me about that. Added implementation in 4826896.